### PR TITLE
v1.9 IPSec key rotation without agent restart

### PR DESCRIPTION
**Pull request by jibi**
This is the v1.9 backport of https://github.com/cilium/cilium/pull/19814.

Just some minor conflicts on `nodediscovery: split StartDiscovery into smaller functions` as the `v1.9` branch does not include the changes to make `NodeDiscovery.LocalNode` a private field

### CODE REVIEW

File: daemon/cmd/daemon.go [hunk 1]
```diff
@@ -690,7 +690,7 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
 	}
 
 	// Must occur after d.allocateIPs(), see GH-14245 and its fix.
-	d.nodeDiscovery.StartDiscovery(nodeTypes.GetName())
+	d.nodeDiscovery.StartDiscovery()
```
Comment (pchaigno) [inline daemon/cmd/daemon.go:694]: Could we backport 9c1cb4856322a80174799dae142ef761e51e2310 instead of cherry-picking a subset of its changes?

**Review by pchaigno — COMMENTED** [review]
I see conflicts reported by GitHub. Is that expected? If it was based on some branch than v1.9, then maybe it could be opened against that branch (e.g., `v1.9.x`)? EDIT: GitHub doesn't allow that :disappointed: 

Other than the comment below and other comments in the upstream PR, looks good to me. Nothing blocking.

File: pkg/nodediscovery/nodediscovery.go [hunk 2]
```diff
@@ -275,6 +289,14 @@ func (n *NodeDiscovery) StartDiscovery(nodeName string) {
 	}
 }
 
+func (n *NodeDiscovery) UpdateLocalNode(fillLocalNode bool) {
```
Comment (gandro) [inline pkg/nodediscovery/nodediscovery.go]: `fillLocalNode` is always true. Why is it needed?

File: pkg/nodediscovery/nodediscovery.go [hunk 3]
```diff
@@ -172,8 +175,44 @@ func (n *NodeDiscovery) JoinCluster(nodeName string) {
 // start configures the local node and starts node discovery. This is called on
 // agent startup to configure the local node based on the configuration options
 // passed to the agent. nodeName is the name to be used in the local agent.
-func (n *NodeDiscovery) StartDiscovery(nodeName string) {
-	n.LocalNode.Name = nodeName
+func (n *NodeDiscovery) StartDiscovery() {
+	n.localNodeLock.Lock()
+	defer n.localNodeLock.Unlock()
```
Comment (gandro) [inline pkg/nodediscovery/nodediscovery.go:179]: Is this mutex needed on v1.9? `StartDiscovery` cannot be called more than once anyway and it doesn't seem to be used anywhere else.

On v1.9, we also cannot really protect writes to `LocalNode`, as it is exported.

**Review by gandro — COMMENTED** [review]
The fsnotify code itself looks good to me. A few comments around the node discovery code though.

File: pkg/nodediscovery/nodediscovery.go (hunk 3 above)
Comment (gandro) [reply pkg/nodediscovery/nodediscovery.go:179]: Oh, I missed the fact that the watcher is calling `fillLocalNode` indirectly. I think that might be inherently racy, as the fields could be written to while some other code is reading it.

**Comment by jibi** [comment]
/test-backport-1.9

**Comment by jibi** [comment]
/test-backport-1.9

**Review by gandro — APPROVED** [review]
Neat, this looks good to me now also in terms of the node discovery logic!

**Comment by jibi** [comment]
/test-backport-1.9

**Pull request closed without merging**